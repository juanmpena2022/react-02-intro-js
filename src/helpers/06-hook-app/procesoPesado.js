export const procesoPesado = (iteraciones) => {
  for (let index = 0; index < iteraciones; index++) {
    console.log("otra iteracion");
  }

  return `Procesados ${iteraciones}`;
};
