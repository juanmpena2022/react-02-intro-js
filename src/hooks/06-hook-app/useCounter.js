import { useState } from "react";

export const useCounter = (initialValue = 10) => {
  const [value, setValue] = useState(initialValue);

  const increment = () => {
    setValue(value + 1);
  };
  const decrement = () => {
    setValue(value - 1);
  };
  const reset = () => {
    setValue(initialValue);
  };

  return {
    value,
    increment,
    decrement,
    reset,
  };
};
