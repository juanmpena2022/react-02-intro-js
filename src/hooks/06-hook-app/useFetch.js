import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
  const isMountedRef = useRef(true);
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });
  useEffect(() => {
    return () => {
      isMountedRef.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        setTimeout(() => {
          if (isMountedRef.current) {
            console.log("se llamo setState");
            setState({
              data: data,
              loading: false,
              error: null,
            });
          } else {
            console.log("no se llamo setState");
          }
        }, 1000);
      })
      .catch((error) => {
        setState({
          data: null,
          loading: false,
          error: error,
        });
      });
  }, [url]);

  return state;
};
