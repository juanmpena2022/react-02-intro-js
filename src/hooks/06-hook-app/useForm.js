import { useState } from "react";

export const useForm = (formStateValue = {}) => {
  const [formState, setFormState] = useState(formStateValue);

  const handleInputChange = ({ target }) => {
    setFormState({
      ...formState,
      [target.name]: target.value,
    });
  };

  return {
    formState,
    handleInputChange,
  };
};
