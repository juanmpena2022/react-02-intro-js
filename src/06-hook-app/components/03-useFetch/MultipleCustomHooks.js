import { useCounter } from "../../../hooks/06-hook-app/useCounter";
import { useFetch } from "../../../hooks/06-hook-app/useFetch";
import "./effect.css";

export const MultipleCustomHooks = () => {
  const { value: counter, increment } = useCounter(1);

  const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;
  const { loading, data } = useFetch(url);
  const { author, quote } = data !== null && data[0];

  return (
    <>
      <h1>MultipleCustomHooks</h1>
      <hr />
      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0">{quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}

      <button className="btn btn-primary" onClick={increment}>Siguiente</button>
    </>
  );
};
