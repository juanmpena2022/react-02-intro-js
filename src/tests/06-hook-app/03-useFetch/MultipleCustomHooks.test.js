import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../06-hook-app/components/03-useFetch/MultipleCustomHooks";
import { useFetch } from "../../../hooks/06-hook-app/useFetch";
import { useCounter } from "../../../hooks/06-hook-app/useCounter";
jest.mock("../../../hooks/06-hook-app/useFetch.js");
jest.mock("../../../hooks/06-hook-app/useCounter.js");

describe("pruebas del componente <MultipleCustomHooks />", () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 10,
      increment: () => {},
      decrement: () => {},
      reset: () => {},
    });
  });

  test("snapshot", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("snapshot debe mostrar la informacion", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "juan",
          quote: "Hola Mundo!!!",
        },
      ],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrar la informacion", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "juan",
          quote: "Hola Mundo!!!",
        },
      ],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find(".mb-0").text()).toBe("Hola Mundo!!!");
  });
});
