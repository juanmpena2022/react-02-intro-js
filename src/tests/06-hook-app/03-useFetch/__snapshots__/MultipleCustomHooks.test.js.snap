// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente <MultipleCustomHooks /> snapshot 1`] = `
<Fragment>
  <h1>
    MultipleCustomHooks
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente
  </button>
</Fragment>
`;

exports[`pruebas del componente <MultipleCustomHooks /> snapshot debe mostrar la informacion 1`] = `
<Fragment>
  <h1>
    MultipleCustomHooks
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
      Hola Mundo!!!
    </p>
    <footer
      className="blockquote-footer"
    >
      juan
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente
  </button>
</Fragment>
`;
